starlette/types.py

```diff
@@ -1,9 +1,18 @@
 import typing
 
+if typing.TYPE_CHECKING:
+    from starlette.applications import Starlette
+
 Scope = typing.MutableMapping[str, typing.Any]
 Message = typing.MutableMapping[str, typing.Any]
 
 Receive = typing.Callable[[], typing.Awaitable[Message]]
 Send = typing.Callable[[Message], typing.Awaitable[None]]
 
 ASGIApp = typing.Callable[[Scope, Receive, Send], typing.Awaitable[None]]
+
+StatelessLifespan = typing.Callable[["Starlette"], typing.AsyncContextManager[None]]
+StatefulLifespan = typing.Callable[
+    ["Starlette"], typing.AsyncContextManager[typing.Mapping[str, typing.Any]]
+]
```

Comment:
Shouldn't it be `ASGIApp` instead of `Starlette`?

Reply:
Well we always pass in a `Starlette` object right? What good does it do to receive a reference to a generic `ASGIApp` object?

Reply:
I was thinking about FastAPI when I asked that, when the `super()` is called on `Router`, they will have an issue over there, I guess? Also, because this module has ASGI related types.

In any case, not that important.

Reply:
Yeah probably need a `self: SelfType, lifespan: Lifespan[SelfType]` or something. Or users can just cast it. Also I don't think there's any FastAPI specific methods they'd want to access from the lifepsan. And with this state thing there's almost no reason to even use the `app` argument anyway...